speaking-portal-project/src/animationFactory/animationProcessor.ts

```diff
@@ -0,0 +1,155 @@
+    return frameData
+}
+
+function willBlink(currentSec: number, frameDur: string, lastBlink: number) {
+    /**
+     * Determines whether or not the avatar will blink based on frame duration and timings
+     *
+     * @remarks
+     * - 1 blink takes ~0.1s, so we only consider frames less than 0.1s long
+     * - Humans typically blink 15-20 times per min, or once every 3-4 seconds
+     * - We want it to be random and not happen exactly once every 3 seconds, so let's say...
+     *      - there's a 40% chance to blink every 4th second
+     *      - there's a 40% chance to blink every 3rd second
+     *      - and don't blink in the same second
+     *
+     * @param currentSec - the current second (integer) of the animation
+     * @param frameDur - the current frame's duration (decimal)
+     * @param lastBlink - the most recent second (integer) in which the avatar blinked
+     *
+     * @returns A boolean
+     *
+    **/
+    let eyes:boolean = false
+    if (parseInt(frameDur) < 0.1 && lastBlink != currentSec) {
+        let n = Math.random()
+        if ((n <= 0.4 && currentSec % 4 == 0) || (n > 0.4 && n <= 0.8 && currentSec % 3 == 0)) {
+            eyes = true
+        }
+    }
+    return eyes
+
+}
+
+function calculateBreathPhase(currentSec: number, lastPoseChange: number, breathPhase: number) {
+    /**
+     * Determines this frame's breath phase value based on time and most recent pose
+     *
+     * @remarks
+     * - A normal respiratory rate in healthy adults is roughly 12 to 20 breaths per minute
+     * - Our avatars' breathing cycle will take 4 seconds, which results in ~15 breaths/min
+     * - Breathing Phases:
+     *      0 = rest
+     *      1 = inhaling or "breathing"
+     *      2 = lungs full
+     *      3 = exhaling or "breathing"
+     *
+     * @returns the new value of breathPhase
+     *
+     **/
+    let result = breathPhase
+    // ensures the idle pose changes occur every second rather than every frame
+    if (lastPoseChange > currentSec) {
+        result += 1
+    }
+    // check to reset breathPhase; values can only be 0 to 4.
+    // TODO: Should this be a type?
+    if (breathPhase == 5) {
+        result = 0
+    }
+    return result
+}
+
+// TODO: if we make breathing a type, we can probably get rid of this function
+function translateBreathFrame(breathPhase: number) {
+    /**
+     * Determines the idle animation frame (inhaling, exhaling, pausing) based on timing and whether the avatar is posing
+     *
+     * @remarks
+     *
+     * @param breathPhase - the current breathing phase of the avatar
+     *
+     * @returns A string to indicate the idle breathing frame
+     *
+     **/
+    if (breathPhase == 0 ){
+        return "_rest" // rest, arms relaxed and down
+    } else if (breathPhase == 1 || breathPhase == 3){
+        return "_breathing" // inhale and exhale (transition between rest and peak)
+    } else { // == 2
+        return "_lungsFull" // lungs are full, arms are out
```

Comment:
This will be dependent on the # of frames we have correct?

I have a sneaking suspicion that if we only have 3 frames, it might look either too fast or too jittery, but once I export the assets I suppose we'll have a better idea - they do look somewhat jittery in photoshop though with just 3 frames

Reply:
No! It's not dependent on the frames! It actually does follow time in seconds! I think it should work okay, but I would like to test it out